File: 0542-01-matrix/0542-01-matrix.py
class Solution:
    def updateMatrix(self, mat: List[List[int]]) -> List[List[int]]:
        M, N = len(mat), len(mat[0])
        q = deque()
        
        res = [[-1] * N for _ in range(M)]
        
        for r in range(M):
            for c in range(N):
                if mat[r][c]==0:
                    res[r][c]=0
                    q.append((r+1,c,1))
                    q.append((r-1,c,1))
                    q.append((r,c+1,1))
                    q.append((r,c-1,1))
        
        while q:
            (r,c,n) = q.popleft()
            if r<0 or r==M or c<0 or c==N:
                continue
            if res[r][c] != -1:
                continue
            
            res[r][c] = n
            q.append((r+1,c,n+1))
            q.append((r-1,c,n+1))
            q.append((r,c+1,n+1))
            q.append((r,c-1,n+1))
        
        return res

File: 0542-01-matrix/0542-01-matrix.js
var updateMatrix = function(mat) {
    const queue = [];

    const res = [];
    res.push(new Array(mat.length))
    for (let row = 0; row < mat.length; row++){
        res[row] = new Array(mat[0].length)
        for (let col = 0; col < mat[0].length; col++){
            
            if (mat[row][col] === 0) {
                res[row][col] = 0;
                queue.push([row, col, 0])
            }
        }
    }

    let directions = [[1,0], [0,1], [-1,0], [0,-1]]
    while (queue.length > 0) {
        const [row, col, level] = queue.shift();

        if (inBounds(mat, row, col)) {

            for (let dir of directions) {
                const newRow = row + dir[0];
                const newCol = col + dir[1];
                if (inBounds(mat, newRow, newCol) && res[newRow][newCol] === undefined) {
                    queue.push([newRow, newCol, level+1])
                    res[newRow][newCol] = level+1
                }
            }
        }
    }

    return res
};

const inBounds = (grid, row, col) => {
    const rowInBounds = 0 <= row && row < grid.length;
    const colInBounds = 0 <= col && col < grid[0].length;
    return rowInBounds && colInBounds
}

